### PR TITLE
Feature/contract api server controller

### PR DESCRIPTION
**Pull request by rhyslbw**
Introduces the `ContractApiServerController`, and concept of _engine client_. Also adds a new test bundle to target the `StubEngineClient`. This will probably get refactored away now we have an interface for the clients. Deferring this as it's not critical

### CODE REVIEW

File: src/server/application/ContractApiServerController.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        resolve(true)
+      })
+    })
+  }
+
+  return {
+    servers,
+    async deploy (contractAddress: Contract['address'], graphQlSchema: IExecutableSchemaDefinition): Promise<boolean> {
+      const allocation = await portManager.getAvailablePort()
+      if (servers.has(contractAddress)) return true
```
Comment (Sam-Jeston) [inline src/server/application/ContractApiServerController.ts:27]: Might as well do the `servers.has` return check prior to getting an available port.

File: src/server/application/ContractApiServerController.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+      const serverInfo = await new ApolloServer({
```
Comment (Sam-Jeston) [inline src/server/application/ContractApiServerController.ts]: I think we should gracefully handle the error here. As much as its unlikely, it is possible that the port you just queried is no longer available by the time the server goes to boot.

File: src/server/application/PortManager.ts [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+      const size = await repository.size()
+      if (size === startingPoolQty) throw new AllPortsAllocated(range)
+      let nextPortNumber
+      if (size === 0) {
+        nextPortNumber = range.lower
+      } else {
+        const last = await repository.getLast()
+        nextPortNumber = last.portNumber + 1
+      }
+      const portAllocation = {
```
Comment (Sam-Jeston) [inline src/server/application/PortManager.ts:28]: Possibly the above is cleaner as

```
const portNumber = size === 0
 ? range.lower
 : (await repository.getLast()).portNumber + 1
```

I prefer assigning to a const through a ternary. But this is subjective, feel free to leave it

File: src/server/core/errors/AllPortsAllocated.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import { CustomError } from 'ts-custom-error'
+import { NumberRange } from '../lib'
+
+export class AllPortsAllocated extends CustomError {
+  constructor (range: NumberRange) {
+    super()
+    this.message = `All ports in the range ${rangeString(range)} have all been allocated`
```
Comment (Sam-Jeston) [inline src/server/core/errors/AllPortsAllocated.ts]:
```suggestion
    this.message = `All ports in the range ${rangeString(range)} have been allocated`
```

File: src/server/infrastructure/engine_clients/StubEngineClient.ts [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import { Bundle, Contract, ContractExecutionInstruction, EngineClient } from '../../core'
+
+export function StubEngineClient (): EngineClient {
```
Comment (Sam-Jeston) [inline src/server/infrastructure/engine_clients/StubEngineClient.ts:3]: 👍

**Review by Sam-Jeston — CHANGES_REQUESTED** [review]

File: src/server/application/PortManager.ts (hunk 3 above)
Comment (rhyslbw) [reply src/server/application/PortManager.ts:28]: Agree, I'll update

File: src/server/application/ContractApiServerController.ts (hunk 2 above)
Comment (rhyslbw) [reply src/server/application/ContractApiServerController.ts]: Good catch

**Review by Sam-Jeston — APPROVED** [review]
👍

**Pull request merged**